File: elixir/nucleotide-count/nucleotide_count.exs
defmodule NucleotideCount do
  @nucleotides [?A, ?C, ?G, ?T]

  @spec count([char], char) :: non_neg_integer
  def count(strand, nucleotide) do
    strand
    |> Enum.count(fn(dna) -> dna == nucleotide end)
  end

  @spec histogram([char]) :: map
  def histogram(strand) do
    @nucleotides
    |> Enum.map(fn(nucleotide) -> {nucleotide, count(strand, nucleotide)} end)
    |> Map.new
  end
end

File: ruby/nucleotide-count/nucleotide_count.rb
class Nucleotide
    NUCLEOTIDES = "ACTG"
    def initialize(str)
        raise ArgumentError unless str.delete(NUCLEOTIDES).empty?
        @strand = str
    end

    def histogram
        NUCLEOTIDES.chars.to_h{|char| [char, @strand.count(char)]}
    end

    def count(letter)
        histogram[letter]
    end

    class << self
        alias_method :from_dna, :new
    end
end
